server/src/main/java/org/elasticsearch/script/SortScript.java

```diff
@@ -0,0 +1,95 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.elasticsearch.script;
+
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.search.Scorer;
+import org.elasticsearch.index.fielddata.ScriptDocValues;
+import org.elasticsearch.search.lookup.LeafSearchLookup;
+import org.elasticsearch.search.lookup.SearchLookup;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.util.Map;
+import java.util.function.DoubleSupplier;
+
+/**
+ * A script used for computing sort values on a per document basis for {@link org.elasticsearch.search.sort.ScriptSortBuilder}.
+ */
+public abstract class SortScript {
+    
+    public static final String[] PARAMETERS = {};
+    
+    private final Map<String, Object> params;
+    private final LeafSearchLookup leafLookup;
+    private DoubleSupplier scoreSupplier = () -> 0.0;
+    
+    public SortScript(Map<String, Object> params, SearchLookup lookup, LeafReaderContext leafContext) {
+        this.params = params;
+        this.leafLookup = lookup.getLeafSearchLookup(leafContext);
+    }
+    
+    // String script sort returns a string
+    public abstract Object execute();
+    
+    // Number script sort returns a double
+    public double executeAsDouble() {
+        return ((Number) execute()).doubleValue();
+    }
+    
+    // TODO: Maybe have two sort script classes? (StringSortScript and NumberScriptSort)
```

Comment:
This is the reason why I labelled this pr as wip. I lean towards having separate classes for string and number based script sorting. This will make the api cleaner, for string based script sorting we can return Object (or maybe String) and for number based script sorting we can return double.